zerver/views/users.py

```diff
@@ -910,6 +915,25 @@ def get_subscription_backend(
     return json_success(request, data=subscription_status)
 
 
+@typed_endpoint
+def get_subscriptions_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    *,
+    user_id: PathOnly[Json[int]],
+) -> HttpResponse:
+    target_user = access_user_by_id(user_profile, user_id, for_admin=False)
+    streams = get_streams_for_user(
+        target_user,
+        include_all=True,
+    )
```

Comment:
This code is completely wrong. Security-wise, it doesn't filter for what the acting user has access to. But it also asks for all channels the user has metadata access to, regardless of whether the user is subscribed!

I think you can perhaps pass that `streams` list into `get_metadata_access_streams`, i.e. `streams = get_metadata_access_streams(get_user_subscribed_streams(target_user), user_group_membership_details)`. 

But it might require a tweak depending whether you want subscriptions to archived channels to appear in the response.

Reply:
The test was set up correctly and wouldn't have failed here because I'm filtering in the next line by if the user is subscribed or not. So (as far as I can tell) it's not completely wrong, but it is inefficient as you pointed out. I can update it to be more efficient and to test query count. Let me know if I'm missing something.